tools/validate.py

```diff
@@ -0,0 +1,138 @@
+            LIMIT %s
+        ) SELECT
+            metric, detail_name, phase, type,
+            AVG(value) as "avg",
+            COALESCE(STDDEV(value), 0) as "stddev",
+            COALESCE(STDDEV(value) / AVG(value), 0) as "rel_stddev",
+            unit
+          FROM phase_stats
+          WHERE
+            phase = %s
+            AND metric IN ($list_replace)
+            AND run_id IN (SELECT id FROM LastXRows)
+          GROUP BY
+            metric, detail_name, phase, type, unit
+    """
+
+    placeholders = ', '.join(['%s'] * len(metrics))
+    query = query.replace('$list_replace', placeholders)
+
+    params = (repo_uri, filename, branch, machine_id, comparison_window, phase, *(metrics))
+    return DB().fetch_all(query=query, params=params)
+
+
+# pylint: disable=broad-except
+if __name__ == '__main__':
+    import argparse
+
+    parser = argparse.ArgumentParser()
+    parser.add_argument('--skip-run', action='store_true', help='Skips running a workload and will just check for STDDEV')
+
+    args = parser.parse_args()
+
+    run_id = None
+    config = GlobalConfig().config
+    try:
+        set_status('measurement_control_start')
+        if not args.skip_run:
+            runner = Runner(
+                name=config['machine']['control_workload']['name'],
+                uri=config['machine']['control_workload']['uri'],
+                uri_type='URL',
+                filename=config['machine']['control_workload']['filename'],
+                branch=config['machine']['control_workload']['branch'],
+                skip_unsafe=True,
+                skip_system_checks=None,
+                full_docker_prune=False,
+                docker_prune=True,
+                job_id=None,
+            )
+            # Start main code. Only URL is allowed for cron jobs
+            run_id = runner.run()
+            build_and_store_phase_stats(run_id, runner._sci)
+
+        set_status('measurement_control_end')
+
+        data = validate_stddev(
+            config['machine']['control_workload']['uri'],
+            config['machine']['control_workload']['filename'],
+            config['machine']['control_workload']['branch'],
+            config['machine']['id'],
+            config['machine']['control_workload']['comparison_window'],
+            config['machine']['control_workload']['phase'],
+            config['machine']['control_workload']['metrics'],
+        )
+
+        print('Validate_stddev returned: ')
+        warning = False
+        info_string_acc = ''
+        for el in data:
+            info_string = f"{el[0]} {el[1]}: {el[4]} +/- {el[5]} {el[6]*100} %"
+            print(info_string)
+            info_string_acc = f"{info_string_acc}\n{info_string}"
+            if el[6] > config['machine']['control_workload']['threshhold']:
+                print(TerminalColors.FAIL, 'Warning. Threshhold exceeded!', TerminalColors.ENDC)
+                warning = True
+        if warning:
```

Comment:
Why are you working with a warning here if you overwrite the `info_string_acc`? Couldn't you just raise right away? Or I would append to the`info_string_ac` so you see where the threshold was exceeded

Reply:
Maybe you missed the line. I am appending: info_string_acc = f"{info_string_acc}\n{info_string}"

Reply:
Totally missed that. Isn't that really inefficient? 
```
didi@marge /tmp % cat a.py
x = ''
for i in range(200_000):
    x = f"{x}{i}"
print(len(x))
didi@marge /tmp % cat b.py
x = []
for i in range(200_000):
    x.append(str(i))
print(len(''.join(x)))
didi@marge /tmp % time python3 a.py
1088890
python3 a.py  1.92s user 0.01s system 99% cpu 1.946 total
didi@marge /tmp % time python3 b.py
1088890
python3 b.py  0.05s user 0.02s system 82% cpu 0.083 total
```

Reply:
wow, I am convinced.

I did this because I saw it as recommendation in https://wiki.python.org/moin/PythonSpeed/PerformanceTips#String_Concatenation . The first one is actally your recommendation, but the second one is an inline replacement. Somehow I thought this is similar.

=> Changed